Wrappers/Python/ccpi/optimisation/algorithms/GradientDescent.py

```diff
@@ -65,29 +75,67 @@ def set_up(self, x_init, objective_function, rate):
             
         self.x = x_init.copy()
         self.objective_function = objective_function
-        self.rate = rate
+        
 
-        self.loss.append(objective_function(x_init))
+        if rate is None:
+            self.k = 0
+            self.update_rate = True
+            self.x_armijo = x_init.copy()
+            # self.rate = self.armijo_rule() * 2
+            # print (self.rate)
+        else:
+            self.rate = rate
+            self.update_rate = False
+        
+        
+        self.update_objective()
         self.iteration = 0
 
-        try:
-            self.memopt = self.objective_function.memopt
-        except AttributeError as ae:
-            self.memopt = False
-        if self.memopt:
-            self.x_update = x_init.copy()
+        self.x_update = x_init.copy()
 
         self.configured = True
         print("{} configured".format(self.__class__.__name__, ))
 
     def update(self):
         '''Single iteration'''
-        if self.memopt:
-            self.objective_function.gradient(self.x, out=self.x_update)
+        
+        self.objective_function.gradient(self.x, out=self.x_update)
+        
+        if self.update_rate:
+            # the next update and solution are calculated within the armijo_rule
+            self.rate = self.armijo_rule() * 2.
+        else:
             self.x_update *= -self.rate
             self.x += self.x_update
-        else:
-            self.x += -self.rate * self.objective_function.gradient(self.x)
+        
+    
 
     def update_objective(self):
         self.loss.append(self.objective_function(self.x))
+
+    def armijo_rule(self):
+        '''Applies the Armijo rule to calculate the step size (rate)
+
+        https://projecteuclid.org/download/pdf_1/euclid.pjm/1102995080
+        '''
+        f_x = self.objective_function(self.x)
+        if not hasattr(self, 'x_update'):
+            self.x_update = self.objective_function.gradient(self.x)
+        while True:
```

Comment:
I don't like this as it may end up in an infinite loop. I'll limit it to a number of iterations which may get configured.

Reply:
I added a rule to evaluate a reasonable number of iterations we want to run, based on `alpha`

basically we want to reduce `alpha` by `2^k` to reach a `rate`

```

rate = alpha / 2^k
```

Given `alpha = 10^gamma` and `rate = 10^-delta`, if `gamma = -delta` this reduces to 

```
kmax = (2 * gamma) / log10(2) = 2 * log10(alpha) / log10(2)
```